pennylane/_queuing_context.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`QueuingContext` abstract base class.
+"""
+import abc
+
+
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for Operations.
+    """
+
+    _active_contexts = []
+    """The list of contexts that are currently active."""
+
+    def __enter__(self):
+        """Adds this instance to the global list of active contexts.
+
+        Returns:
+            QueuingContext: This instance
+        """
+        QueuingContext._active_contexts.append(self)
+
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        """Remove this instance from the global list of active contexts.
+        """
+        QueuingContext._active_contexts.remove(self)
```

Comment:
Is it possible for the same QNode to be added to the context list twice? In which case, will this remove method potentially remove the wrong QNode?

Reply:
Only in the core code, because we have the QNode use the context manager template
```with self:
        ... all operations created here go into the qnode ...
```
For a second QNode to go active, one would have to _construct a qnode inside the with block of the qnode (which is also in _construct).

And for the same QNode, this is basically not possible currently.